.github/workflows/validate.yml

```diff
@@ -17,6 +17,7 @@ jobs:
     strategy:
       matrix:
         node: [12.13, 12, 14, 16]
+        react: [^16.9.0, ^17.0.2, next]
```

Comment:
Can we make this `[16.9.0, ^16, ^17, next]` so we ensure it works with our minimum version, latest for each supported version and the next version?

Reply:
I'm not sure why the latest test run failed on CI

Reply:
It's failing coverage and then there's a thing to cancel running jobs when on job fails

```txt
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] File                    | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] All files               |   88.26 |    89.74 |   85.71 |   87.67 |                   
[test]  src                    |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |     100 |      100 |     100 |     100 |                   
[test]  src/core               |     100 |      100 |     100 |     100 |                   
[test]   asyncUtils.ts         |     100 |      100 |     100 |     100 |                   
[test]   cleanup.ts            |     100 |      100 |     100 |     100 |                   
[test]   console.ts            |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]  src/dom                |   63.64 |       50 |   53.85 |   63.64 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |      60 |       50 |   53.85 |      60 | 20-34             
[test]  src/helpers            |     100 |      100 |     100 |     100 |                   
[test]   createTestHarness.tsx |     100 |      100 |     100 |     100 |                   
[test]   error.ts              |     100 |      100 |     100 |     100 |                   
[test]   promises.ts           |     100 |      100 |     100 |     100 |                   
[test]  src/native             |     100 |      100 |     100 |     100 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |     100 |      100 |     100 |     100 |                   
[test]  src/server             |   57.78 |       50 |   52.94 |   57.78 |                   
[test]   index.ts              |     100 |      100 |     100 |     100 |                   
[test]   pure.ts               |   55.81 |       50 |   52.94 |   55.81 | 23-57             
[test]  src/types              |       0 |        0 |       0 |       0 |                   
[test]   index.ts              |       0 |        0 |       0 |       0 |                   
[test]   react.ts              |       0 |        0 |       0 |       0 |                   
[test] ------------------------|---------|----------|---------|---------|-------------------
[test] Jest: "global" coverage threshold for statements (100%) not met: 88.26%
[test] Jest: "global" coverage threshold for branches (100%) not met: 89.74%
[test] Jest: "global" coverage threshold for lines (100%) not met: 87.67%
[test] Jest: "global" coverage threshold for functions (100%) not met: 85.71%
[test] 
[test] Test Suites: 25 passed, 25 total
[test] Tests:       298 passed, 298 total
[test] Snapshots:   0 total
[test] Time:        30.323 s
[test] Ran all test suites.
[test] npm run test --silent -- --coverage exited with code 1
Error: Process completed with exit code 1.
```

Reply:
Hmm... This is interesting because we can't cover both sides in a single run, but we will cover all sides across all the jobs

Reply:
ah, I see it, hmm i'm not sure how we'd solve for that because we won't cover those lines depending on the React version we're using

Reply:
Drop the coverage threshold to 80% for now. I've got an idea, but will need to get some time in front of my editor to work it out.

Reply:
heh, got unlucky since the `next` tests don't pass right now that they canceled all the others, maybe those can be changed to avoid canceling the rest

Reply:
That would be ideal if the `next` tests are allowed to fail, but the others are strict.

Reply:
I've done what was I was thinking of doing (`createLegacyRoot`). It reduced the amount of code that is uncovered depending on which react version is used, to the point where I'm fairly comfortable ignoring those lines as the logic is quite light around it.

I also introduced the `react/next` and `react-dom/next` types, however this broke the `node` and `jest` types ([and any other global types](https://www.typescriptlang.org/tsconfig#types) which I'm not thrilled with the fix for (also explicitly including those types).  If anyone knows a better way of adding those types them please let me know, otherwise this if fine for now.

I haven't looked into why the the async (are related) tests are failing with react 18 yet.